### PR TITLE
[CARBONDATA-3837] Fallback to the original plan when mv rewrite throw exception

### PR DESCRIPTION
**Pull request by QiangCai**
### Why is this PR needed?
 1. all plans are checking MVRewriteRule, 
 2. if MVRewriteRule throw an exception,  it will lead to query failure
 
 ### What changes were proposed in this PR?
1. only the query should check MVRewriteRule, other plans should skip it quickly.

2. catch all exceptions of MVRewriteRule, and fallback to original plan.
    
 ### Does this PR introduce any user interface change?
 - No

 ### Is any new testcase added?
 - Yes

### CODE REVIEW

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3080/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1359/

**Comment by jackylk** [comment]
retest this please

File: integration/spark/src/main/spark2.4/org/apache/spark/sql/CarbonToSparkAdapter.scala [hunk 1]
```diff
@@ -193,9 +193,17 @@ class CarbonOptimizer(
     catalog: SessionCatalog,
     optimizer: Optimizer) extends Optimizer(catalog) {
 
-  private lazy val mvRules = Seq(Batch("Materialized View Optimizers", Once,
+  private lazy val _mvRules = Seq(Batch("Materialized View Optimizers", Once,
     Seq(new MVRewriteRule(session)): _*))
 
+  private def mvRules: Seq[Batch] = {
+    // enable mv by default
+    session.conf.get("spark.carbon.mv.enable", "true").toBoolean match {
```
Comment (jackylk) [inline integration/spark/src/main/spark2.4/org/apache/spark/sql/CarbonToSparkAdapter.scala:201]: make a  constant for default value

File: integration/spark/src/main/scala/org/apache/spark/sql/optimizer/MVRewriteRule.scala [hunk 2]
```diff
@@ -39,15 +39,31 @@ import org.apache.carbondata.view.MVFunctions.DUMMY_FUNCTION
...
+      case _: Command => return logicalPlan
+      case _: LocalRelation => return logicalPlan
+      case _ =>
+    }
+    try {
+      tryRewritePlan(logicalPlan)
+    } catch {
+      case e =>
+        // catch all exceptions to avoid to impact query.
+        // if MVRewriteRule throw exception, here will fallback to original plan.
```
Comment (jackylk) [inline integration/spark/src/main/scala/org/apache/spark/sql/optimizer/MVRewriteRule.scala:60]:
```suggestion
        // if exception is thrown while rewriting the query, will fallback to original query plan.
```

File: integration/spark/src/main/scala/org/apache/spark/sql/optimizer/MVRewriteRule.scala [hunk 3]
```diff
@@ -39,15 +39,31 @@ import org.apache.carbondata.view.MVFunctions.DUMMY_FUNCTION
...
+    logicalPlan match {
+      case _: Command => return logicalPlan
+      case _: LocalRelation => return logicalPlan
+      case _ =>
+    }
+    try {
+      tryRewritePlan(logicalPlan)
+    } catch {
+      case e =>
+        // catch all exceptions to avoid to impact query.
```
Comment (jackylk) [inline integration/spark/src/main/scala/org/apache/spark/sql/optimizer/MVRewriteRule.scala:59]: This line can be deleted

File: integration/spark/src/main/spark2.4/org/apache/spark/sql/CarbonToSparkAdapter.scala (hunk 1 above)
Comment (QiangCai) [reply integration/spark/src/main/spark2.4/org/apache/spark/sql/CarbonToSparkAdapter.scala:201]: this is a runtime configuration.
so the user can use the SET command to change it.

File: integration/spark/src/main/spark2.3/org/apache/spark/sql/CarbonToSparkAdapter.scala [hunk 4]
```diff
@@ -154,9 +154,17 @@ class CarbonOptimizer(
     catalog: SessionCatalog,
     optimizer: Optimizer) extends Optimizer(catalog) {
 
-  private lazy val mvRules = Seq(Batch("Materialized View Optimizers", Once,
+  private lazy val _mvRules = Seq(Batch("Materialized View Optimizers", Once,
     Seq(new MVRewriteRule(session)): _*))
 
+  private def mvRules: Seq[Batch] = {
+    // enable mv by default
+    session.conf.get("spark.carbon.mv.enable", "true").toBoolean match {
```
Comment (ajantha-bhat) [inline integration/spark/src/main/spark2.3/org/apache/spark/sql/CarbonToSparkAdapter.scala:162]: How user knows this property exist ? Need to add it in some document ?
And why not as carbon property ?

File: integration/spark/src/main/spark2.4/org/apache/spark/sql/CarbonToSparkAdapter.scala (hunk 1 above)
Comment (jackylk) [reply integration/spark/src/main/spark2.4/org/apache/spark/sql/CarbonToSparkAdapter.scala:201]: conf name should start with carbon, and please add in doc also

File: integration/spark/src/main/spark2.3/org/apache/spark/sql/CarbonToSparkAdapter.scala (hunk 4 above)
Comment (jackylk) [reply integration/spark/src/main/spark2.3/org/apache/spark/sql/CarbonToSparkAdapter.scala:162]: please add in https://github.com/apache/carbondata/blob/master/docs/configuration-parameters.md 
conf name should start with carbon, changed to `carbon.mv.enabled`

Comment (QiangCai) [reply integration/spark/src/main/spark2.3/org/apache/spark/sql/CarbonToSparkAdapter.scala]: fixed

File: core/src/main/java/org/apache/carbondata/core/constants/CarbonCommonConstants.java [hunk 5]
```diff
@@ -1562,6 +1562,20 @@ private CarbonCommonConstants() {
...
   public static final String CARBON_LUCENE_INDEX_STOP_WORDS_DEFAULT = "false";
 
+  //////////////////////////////////////////////////////////////////////////////////////////
+  // MV parameter start here
+  //////////////////////////////////////////////////////////////////////////////////////////
+
+  /**
+   * Property to enable MV rewrite
+   */
+  @CarbonProperty(dynamicConfigurable = true)
```
Comment (ajantha-bhat) [inline core/src/main/java/org/apache/carbondata/core/constants/CarbonCommonConstants.java:1572]: You wanted to make this dynamic configurable, but forgot to add in 
SessionParams#validateKeyValue ?

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3112/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1388/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3113/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1389/

**Comment by jackylk** [comment]
LGTM

**Pull request closed without merging**